File: src/clu/sources/os_release.py
import logging

from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts
from clu.sources import Source
from clu.input import text_file
from clu.sources import PARSE_FAIL_MSG

log = logging.getLogger(__name__)


class OsRelease(Source):
    def provides(self, provides: Provides) -> None:
        provides["os.distro.name"] = self
        provides["os.distro.version"] = self

    def requires(self, requires: Requires) -> None:
        requires.files.append("/etc/os-release")

    def parse(self, facts: Facts) -> None:
        data = text_file("/etc/os-release")
        log.debug(f"{data=}")
        if not data:
            facts["os.distro.name"] = PARSE_FAIL_MSG
            facts["os.distro.version"] = PARSE_FAIL_MSG
            return

        for line in data.splitlines():
            if "=" not in line:
                continue
            key, value = line.split("=", 1)
            key = key.strip().strip('"')
            value = value.strip().strip('"')
            log.debug(f"{key=}")
            log.debug(f"{value=}")

            if key == "ID":
                facts["os.distro.name"] = value
            elif key == "VERSION_ID":
                facts["os.distro.version"] = value

File: pkg/facts/sources/os_release.go
package sources

// Go port of src/clu/sources/os_release.py
// Extracts os.distro.name (ID) and os.distro.version (VERSION_ID) from /etc/os-release.

import (
	"strings"

	"github.com/huntermatthews/clu/pkg/facts/types"
	"github.com/huntermatthews/clu/pkg/input"
)

// OsRelease parses the /etc/os-release file for distro identification.
type OsRelease struct{}

var osReleaseFacts = map[string]*types.Fact{
	"os.distro.name":    {Name: "os.distro.name", Tier: types.TierOne},
	"os.distro.version": {Name: "os.distro.version", Tier: types.TierOne},
}

// Provides registers distro fact keys.
func (o *OsRelease) Provides(p types.Provides) {
	for name := range osReleaseFacts {
		p[name] = o
	}
}

// Requires declares the file dependency.
func (o *OsRelease) Requires(r *types.Requires) { r.Files = append(r.Files, "/etc/os-release") }

// Parse reads /etc/os-release, setting types.ParseFailMsg on failure. Only ID and VERSION_ID are used.
func (o *OsRelease) Parse(f *types.FactDB) {
	data, err := input.FileReader("/etc/os-release")
	if err != nil || data == "" { // treat empty/error uniformly
		for _, fact := range osReleaseFacts {
			fact.Value = types.ParseFailMsg
			f.AddFact(*fact)
		}
		return
	}

	var id, version string
	for _, line := range strings.Split(data, "\n") {
		if !strings.Contains(line, "=") {
			continue
		}
		parts := strings.SplitN(line, "=", 2)
		key := strings.Trim(parts[0], " \"")
		value := strings.Trim(parts[1], " \"")
		switch key {
		case "ID":
			id = value
		case "VERSION_ID":
			version = value
		}
	}
    // BUG: this is hot garbage
	if id == "" {
		id = types.ParseFailMsg
	}
	if version == "" {
		version = types.ParseFailMsg
	}
	osReleaseFacts["os.distro.name"].Value = id
	osReleaseFacts["os.distro.version"].Value = version

	// Add all facts to the FactDB
	for _, fact := range osReleaseFacts {
		f.AddFact(*fact)
	}
}
